compiler/optimizations/inferConstRefs.cpp

```diff
@@ -366,43 +316,263 @@ static bool inferConstRef(Symbol* sym) {
+    }
+  } else if (!retval && info) {
+    // Not the first call, but something happened that prevents us from marking
+    // this symbol as const. Now that we know, indicate that this is finalized.
+    info->finalizedConstness = true;
+  }
+
+  return retval;
+}
+
+static bool inferRefToConst(Symbol* sym) {
+  if (!sym->isRef()) return false;
+
+  bool isConstRef = sym->qualType().getQual() == QUAL_CONST_REF;
+  const bool wasRefToConst = sym->hasFlag(FLAG_REF_TO_CONST);
+
+  ConstInfo* info = infoMap[sym];
+
+  // If this ref isn't const, then it can't point to a const thing
+  if ((info && info->finalizedRefToConst) || wasRefToConst || !isConstRef) {
+    return wasRefToConst;
+  }
+
+  bool isFirstCall = false;
+  if (info && info->alreadyCalled == false) {
+    isFirstCall = true;
+    info->alreadyCalled = true;
+  }
+
+  bool retval = true;
+
+  if (isArgSymbol(sym)) {
+    // Check each call and set retval to false if any actual is not a ref
+    // to a const.
+    FnSymbol* fn = toFnSymbol(sym->defPoint->parentSymbol);
+    if (fn->hasFlag(FLAG_VIRTUAL)) {
+      // Not sure how to best handle virtual calls, so simply set false for now
+      retval = false;
+    } else {
+      if (isFirstCall && info) {
+        INT_ASSERT(info->fnUses == NULL);
+        info->fnUses = fn->firstSymExpr();
+      } else if (info == NULL) {
+        retval = false;
+      }
+      // Need this part to be re-entrant in case of recursive functions
+      while (info && info->fnUses != NULL && retval) {
+        SymExpr* se = info->fnUses;
+        info->fnUses = se ? se->symbolSymExprsNext : NULL;
+
+        CallExpr* call = toCallExpr(se->parentExpr);
+        INT_ASSERT(call && call->isResolved());
+
+        Symbol* actual = toSymExpr(formal_to_actual(call, sym))->symbol();
+
+        if (actual->isRef()) {
+          // I don't think we technically need to skip if the actual is the
+          // same symbol as the formal, but it makes things simpler.
+          if (actual != sym && !inferRefToConst(actual)) {
+            retval = false;
+          }
+        } else {
+          // Passing a non-ref actual to a reference formal is currently
+          // considered to be the same as an addr-of
+          if (actual->qualType().getQual() != QUAL_CONST_VAL) {
+            retval = false;
+          }
+        }
+      }
+    }
+  }
+
+  while (info && info->hasMore() && retval) {
+    SymExpr* use = info->next();
+
+    CallExpr* call = toCallExpr(use->parentExpr);
+    if (call == NULL) continue;
+
+    if (isMoveOrAssign(call)) {
+      if (use == call->get(1)) {
+        if (SymExpr* se = toSymExpr(call->get(2))) {
+          if (se->isRef() && !inferRefToConst(se->symbol())) {
+            retval = false;
+          }
+        }
+        else {
+          CallExpr* RHS = toCallExpr(call->get(2));
+          INT_ASSERT(RHS);
+          if (RHS->isPrimitive(PRIM_ADDR_OF) ||
+              RHS->isPrimitive(PRIM_SET_REFERENCE)) {
```

Comment:
We're looking for this kind of pattern:

PRIM_MOVE use, <expr>

'use' should always be call->get(1). The branch with addr-of and set-reference will be entered if the RHS of the move is not a SymExpr.